src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1839,9 +1844,162 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+                    _ => {
+                        return None;
+                    }
+                };
+
+                debug!(
+                    "try_note_static_dyn_trait_impl: call_expr={:?} call_expr.kind={:?}",
+                    call_expr, call_expr.kind,
+                );
+
+                let tables = self.in_progress_tables.unwrap().borrow();
+
+                if let hir::ExprKind::MethodCall(_, _, args) = &call_expr.kind {
+                    let method_def_id = tables.type_dependent_def_id(call_expr.hir_id).unwrap();
+                    let trait_def_id = self.tcx.trait_of_item(method_def_id);
+
+                    debug!("try_note_static_dyn_trait_impl: trait_def_id={:?}", trait_def_id);
+
+                    // As this is a method call expression, we have at least one argument.
+                    let receiver_arg = &args[0];
+
+                    debug!(
+                        "try_note_static_dyn_trait_impl: receiver_arg.kind={:?}",
+                        tables.expr_ty(receiver_arg).kind
+                    );
+
+                    if let ty::Ref(_, arg_ty, _) = tables.expr_ty(receiver_arg).kind {
+                        let mut trait_impl = None;
+
+                        self.tcx.for_each_relevant_impl(
+                            trait_def_id.unwrap(),
+                            arg_ty,
+                            |impl_def_id| {
+                                trait_impl = Some(impl_def_id);
+                            });
+
+                        if let Some(impl_def_id) = trait_impl {
+                            self.note_dyn_impl_and_suggest_anon_lifetime(
+                                err,
+                                impl_def_id,
+                                &dyn_trait_name
+                            );
+                            return Some(());
+                        }
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    pub fn note_dyn_impl_and_suggest_anon_lifetime(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        impl_def_id: DefId,
+        dyn_trait_name: &str,
+    ) {
+        let impl_span = self.tcx.sess.source_map()
+            .def_span(self.tcx.def_span(impl_def_id));
+        debug!(
+            "try_report_static_dyn_trait: impl_span={:?}", impl_span,
+        );
+
+        err.span_note(
+            impl_span,
+            &format!("...because this implementation requires it"),
+        );
+
+        if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(impl_span) {
+            err.span_suggestion(
+                impl_span,
+                &format!(
+                    "you can add an explicit constraint to the implementation so that it applies \
+                    to types with less than `'static` lifetime",
+                ),
+                snippet.replace(dyn_trait_name, &format!("{} + '_", &dyn_trait_name)),
```

Comment:
I don't have time to do an in-depth review right now, but we should try to come up with a different way of doing this. Fiddling with the textual representation of the code is a recipe for pain later on. It produces output that isn't always correct (like the problem with the missing parens and some incorrect suggestions I noticed below), and is brittle and affected by whitespace. I'll take another look later to see if I can have more actionable feedback on what to do instead.